game-core/src/main/java/games/strategy/engine/framework/GameState.java

```diff
@@ -0,0 +1,18 @@
+package games.strategy.engine.framework;
+
+/**
+ * Holder for static 'game started' state.
+ */
+public enum GameState {
+  ;
+
+  private static boolean started = false;
```

Comment:
@ssoloff what are your thoughts? Do you agree the usage of `System#get/setProperty()` is already not thread safe?

Reply:
> Do you agree the usage of System#get/setProperty() is already not thread safe?

To be sure, the usage from in the context of our code. We call the two methods about 7 lines apart from each other. That pattern is itself not thread safe, considering two threads executing in lock step we see there would be a problem.

Reply:
Sorry for not being clear.  My concern isn't _concurrent_ access to this field by multiple threads.  Rather, it is that a change in value is not safely _published_ across threads.  Consider the following scenario:

1. Thread 1 calls `ServerModel#getServerProps()` and subsequently calls `GameState#setStarted()`.
1. At a later time, Thread 2 calls `ServerModel#getServerProps()` and subsequently calls `GameState#notStarted()`.

Without a memory barrier, Thread 2 is only guaranteed to see `started` == `true` _eventually_, which may not be when step (2) runs above.  `volatile` introduces a memory barrier during a read/write so that a read that _happens after_ a write is guaranteed to see the current value in memory rather than a stale value in a processor cache.  That's the same guarantee the use of `Properties#get/setProperty()` provides (although it is accomplished via `synchronize`).

(In practice, you typically only see such unsafe publication cause issues when your system has multiple _physical processors_, not just multiple _cores_.  Typically because cores on the same processor share the same L2 cache.)

Out of curiosity, I added a trace to `ServerModel#getServerProps()` to see which threads pass through it.  Indeed, it appears every time you click "Host Networked Game" from the main menu, this code runs on a different thread:

```
ServerModel#getServerProps: Thread[Thread-10,6,main]
ServerModel#getServerProps: Thread[Thread-11,6,main]
ServerModel#getServerProps: Thread[Thread-12,6,main]
```

I agree that the API (either using `GameState` or `get/setProperty()`) is not completely thread-safe without external synchronization because it could lead to a race condition if multiple threads attempt to read/write this field concurrently.  However, the `get/setProperty()` implementation at least guarantees safe publication, whereas the `GameState` implementation does not.  This is the "loss of parity" I was talking about in my original comment.

(If concurrent, race-free access were required, a better solution would be to modify the API to have a single compare-and-set method rather than two get/set methods.  That would better match the current use cases where we always do something like this:

```java
if (GameState.notStarted()) {
  // ... do something ...
  GameState.setStarted();
}
```

For example:

```java
if (GameState.setStartedIfNotStarted()) {
  // ... do something ...
}
```

But that's out of scope and, as pointed out above, probably unnecessary at this time.)

> I don't think we are in the habit of documenting thread unsafe code, I believe that is the assumption unless something is explicitly called out as 'thread-safe'.

In general, I think it's pretty standard that _instance_ methods are assumed to not be thread-safe unless otherwise documented.  However, _static_ methods are typically assumed to be thread-safe unless otherwise documented.

Reply:
Indeed, understood about the write barrier and two threads seeing updates by one another. But, if we consider two threads executing this code concurrently, it's broken, so getting atomic at the point of the check but not set is far from sufficient for thread safety.

I believe your last example is the one that should be used if we were to make this thread safe.

In your experimentation, were the new threads always the result of a user driven action? In those cases, there is some potential to limit the concurrency just based on how many new threads the user can spawn.